Refactor mount fakes

codenrhoden opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup


**What this PR does / why we need it**:
As part of the review for https://github.com/kubernetes/utils/pull/100, it was requested to put all fakes in a `testing` subpackage.

After discussion, a different package structures was agreed on with @saad-ali and @thockin, see https://github.com/kubernetes/kubernetes/pull/81423#pullrequestreview-278070861.

This patch moves fake.go to mount_fake.go, and follows to principle of
always returning a discrete type rather than an Interface. All callers
of "FakeMounter" are changed to instead use "NewFakeMounter()". The
FakeMounter "Log" struct member is changed to not be exported, and
instead only access through a new "GetLog()" method.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:


**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

codenrhoden commented:
OMG. Just realized, thanks to this [test result](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/81423/pull-kubernetes-verify/1161746463441555457) that moving to a `types` package resolves our issue with the golinter! This means that the package now is all good from the linting perspective, negating all the issues raised in https://github.com/kubernetes/kubernetes/pull/81186.  🎉

mattjmcnaughton reviewed (commented):
Mind addressing the govet failures that are causing `pull-kubernetes-verify` to fail.

/test pull-kubernetes-bazel-test

Since this is such a big diff, I imagine we'll rely pretty heavily on CI for validation, so going to wait for that to be green before reviewing.

Also, to confirm - this pr changes no functionality correct? It just changes the modules of certain functionality?

msau42 commented:
@kubernetes/sig-storage-pr-reviews 
/assign

codenrhoden commented:
> Mind addressing the govet failures that are causing `pull-kubernetes-verify` to fail.

yep, of course. Should be done now.

> Also, to confirm - this pr changes no functionality correct? It just changes the modules of certain functionality?

That's the idea. The only real change is that in order for the `FakeMounter` to be independent in the `testing` package, it can't call into shared functions found in the top level `mount` package. There were two places that did this. They were simple pieces of re-used logic, and I merely copied them over. So - some logic is copied. That's probably worth pointing out. Doesn't change the "real" implementation, only the fake one.

codenrhoden commented:
/test pull-kubernetes-e2e-gce-100-performance

codenrhoden commented:
/test pull-kubernetes-e2e-gce-csi-serial

codenrhoden commented:
/retest

saad-ali commented:
/assign

mattjmcnaughton commented on pkg/kubelet/kubelet.go:
```diff
@@ -111,15 +115,13 @@ import (
 	"k8s.io/kubernetes/pkg/volume"
 	"k8s.io/kubernetes/pkg/volume/csi"
 	"k8s.io/kubernetes/pkg/volume/util/subpath"
-	utilexec "k8s.io/utils/exec"
```
Curious how this change relates to your pr?

codenrhoden replied:
You are right that it doesn't. I tried to describe why when I opened the PR:

> Since I touched a lot of files outside of mount to use the new imports, and I was editing imports, I also followed the goimports standard of putting imports in order of: (1) go std libs (2) 3rd party (3) k8s.io 3rd party (4) k8s.io/kubernetes. The lack of consistency around this was getting to me.

I understand, though, if people want me to change it back. Part of it for me is that I have been coming at this from a mindset of moving a package **out** of `k/k`, and with that mind, it's been really useful to glance at a file and see if there are any `k/k` imports... When the imports aren't correctly grouped, it's hard to tell that. But that's just me being pedantic.

mattjmcnaughton reviewed (approved):
/lgtm

conceptually, this makes sense to me from a sig-node perspective. definitely would love other reviewers to weigh in on this.

mattjmcnaughton commented:
On Fri, Aug 16, 2019 at 09:37:21AM -0700, Travis Rhoden wrote:
> codenrhoden commented on this pull request.
> 
> 
> 
> >  	nodeutil "k8s.io/kubernetes/pkg/util/node"
>  	"k8s.io/kubernetes/pkg/util/oom"
>  	"k8s.io/kubernetes/pkg/util/selinux"
>  	"k8s.io/kubernetes/pkg/volume"
>  	"k8s.io/kubernetes/pkg/volume/csi"
>  	"k8s.io/kubernetes/pkg/volume/util/subpath"
> -	utilexec "k8s.io/utils/exec"
> 
> You are right that it doesn't. I tried to describe why when I opened the PR:
> 
> > Since I touched a lot of files outside of mount to use the new imports, and I was editing imports, I also followed the goimports standard of putting imports in order of: (1) go std libs (2) 3rd party (3) k8s.io 3rd party (4) k8s.io/kubernetes. The lack of consistency around this was getting to me.
> 
> I understand, though, if people want me to change it back. Part of it for me is that I have been coming at this from a mindset of moving a package **out** of `k/k`, and with that mind, it's been really useful to glance at a file and see if there are any `k/k` imports... When the imports aren't correctly grouped, it's hard to tell that. But that's just me being pedantic.
> 
> -- 
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub:
> https://github.com/kubernetes/kubernetes/pull/81423#discussion_r314798661

How difficult would it be to change it back? I don't feel super strong... but as
a general rule, its usually good to keep PRs as small as possible, and this one
is already a pretty big refactor. So if there's optional components that we
could push to a later PR in order to decrease the size of this PR, I'd
be in favor of that.

codenrhoden commented:
> How difficult would it be to change it back? I don't feel super strong... but as a general rule, its usually good to keep PRs as small as possible, and this one is already a pretty big refactor. So if there's optional components that we could push to a later PR in order to decrease the size of this PR, I'd be in favor of that.

Okay, I've changed a lot of them back. In some places it made sense to keep it, but I removed all the ones that seemed to come out of nowhere. It was a useful exercise to go through them all, because in an earlier version of this PR, I had files that were changed and then made some simplifications that made those changes no longer necessary. But I ended up with several files (12 to be exact) that had **only* import statement changes, and no code changes. Those have all been removed now, making this PR a tad bit smaller as well.

changed lines went from `+1,248 −908` to `+1,158 −853`

codenrhoden commented:
/cc @hogepodge

codenrhoden commented:
/retest

saad-ali commented on pkg/util/mount/types/types.go:
```diff
@@ -0,0 +1,112 @@
+limitations under the License.
+*/
+
+package types
```
I really don't like this pattern of stuffing all interfaces in to a single package.

A much cleaner pattern is:

- github.com/bar/foo
  - foo.go
    - `type Foo interface` - the interface that all external references should depend on
    - `func NewFoo(...) Foo` - returns a concrete implementation of interface Foo
  - foo_test.go
    - Tests the concrete implementation of Foo
  - fake_foo.go
    - `func NewFakeFoo(...) Foo` - returns a fake implementation of interface Foo

This makes it dead simple to use, if you want foo (real or a fake), import "github.com/bar/foo".

Looking at the comment that spurred this refactor (https://github.com/kubernetes/utils/pull/100#discussion_r311318715), the problem there was that the package k8s.io/utils/exec doesn't implement a fake, so a thin wrapper `mounter/exec.go` was implemented. The fix for that is not to stuff all the types in to a single file. We should look in to if we really need the wrapper "mount/exec" Interface around "k8s.io/utils/exec"? Or can we use "k8s.io/utils/exec" directly? You could create a package just for the exec fake that implements a fake for the interface in utils/exec without creating a wrapper for the non-test code OR even better, put the fake in the utils/exec package.

saad-ali replied:
Also I realize this is a big change, after all the heroic work you've done. So please feel free to push back. My thinking is if we're gonna refactor this, let's refactor it right.

codenrhoden replied:
I hear you. And FWIW, I agree with your package structure. Either direction we go is a big change. I just want to make sure we have an agreed upon approach that will also satisfy @thockin desires before putting in more effort.

There are also "fake" implementations for the mounter and the HostUtils interface. Whether those are in a `testing` subpackage or not is not super important to me (but like I said, I have a slight preference for your approach, Saad), I just want to make sure that whatever we produce will be approved at a higher level.

I'm in full agreement with "let's refactor it right". This is a good opportunity to clean up a lot of things. My only concern is getting it in for 1.16. I know a lot of interested parties are hoping for that. The amount of code these changes touch is massive, which leads to difficulty getting `/approve`'s. There are definitely tradeoffs. But as long as it's possible, let's do it right.

hogepodge replied:
The ugliness that our patch depends on, vendoring utils/mount in staging/openstack, is by design a short-lived hack and will be removed almost immediately after 1.16 release. While this refactor would be a "nice-to-have" for us, I don't think it's a "must-have".

codenrhoden replied:
@saad-ali Any agreement between you and Tim on preferred package structure here?

saad-ali replied:
I was able to sync up with @thockin. He pointed me to the following go lang principles.
1) "Accept interfaces, return concrete types" (see [1](http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/) and [2](https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8)).
2) "Consumer should define the interface" (see [1](https://github.com/golang/go/wiki/CodeReviewComments#interfaces) and [2](https://twitter.com/davecheney/status/942593128355192832?s=20)).

"Accept interfaces" makes perfect sense.

"Return concrete types" and "consumer should define the interface" were new to me. The argument for doing so is that "defining an interface upfront is usually a code smell for overengineering" and it not doing so "reduces dependencies" (see [1](https://blog.chewxy.com/2018/03/18/golang-interfaces/) and [2](https://news.ycombinator.com/item?id=16623720)).

In this case, given that we have multiple concrete implementations of the mounter: real (linux, windows) and fake. And we our external consumers will want to use both our real and fake implementations, we have sufficient justification to define an interface for Mounter and export fake and real concrete implementations.

Therefore, as far as the file structure is concerned, for the top level mounter package, we should do what I recommended above ("Proposal 1") with a small modification, the return type for `NewFoo(...)` and `NewFakeFoo(...)` should be the concrete type not an interface.

For sub-packages that mounter will depend on, you can follow the same pattern as long as there are multiple consumers of both real and fake types. If not, then we should follow go lang guidance.

codenrhoden replied:
Thanks for the clarification, and the links for reference.

I will refactor this PR after https://github.com/kubernetes/kubernetes/pull/81794 goes in, as that will also reduce the scope of what's happening. In fact, I already have a follow-on commit to that one that does most of what we need for `HostUtils`, but not yet for `Mount`.

Progress!

saad-ali replied:
Sounds good, thanks a lot for your patience and hard work!!

codenrhoden committed:
copy exec file to preserve history

codenrhoden committed:
move exec files back

codenrhoden committed:
Move mount/fake.go to mount/fake_mount.go

This patch moves fake.go to mount_fake.go, and follows to principle of
always returning a discrete type rather than an Interface. All callers
of "FakeMounter" are changed to instead use "NewFakeMounter()". The
FakeMounter "Log" struct member is changed to not be exported, and
instead only access through a new "GetLog()" method.

msau42 commented:
@codenrhoden was this PR waiting on something else?

codenrhoden commented:
> @codenrhoden was this PR waiting on something else?

Nope, looks like I just need to do the rebase and this could go in any time. The `exec` related stuff is going to be not-needed soon, but will keep it in here just for completeness.

codenrhoden commented:
@saad-ali Can you take a new look at this one? It should be good to go now.

/assign @misterikkit 
/assign @derekwaynecarr @Random-Liu

saad-ali commented:
/assign

codenrhoden commented:
/retest

codenrhoden commented:
/retest

codenrhoden commented:
/retest

codenrhoden commented:
/retest

codenrhoden commented:
/test pull-kubernetes-e2e-gce-storage-slow

codenrhoden commented:
/retest

jsafrane commented:
/lgtm
/approve

codenrhoden commented:
/assign @Random-Liu @derekwaynecarr @tallclair 
Any chance I can get a kubelet approver?

dims commented:
/milestone v1.17

derekwaynecarr commented:
the kubelet changes all look like straight forward refactors.

thanks!

/approve
/lgtm

vladimirvivien commented:
/lgtm

codenrhoden commented:
Rebased after a new merge conflict in the same PR batch. Needed to add this line to use the new struct member added in a different PR: https://github.com/kubernetes/kubernetes/pull/81423/files#diff-b962320d780ac74dffce9da68eda4f4cR616
